contracts/treasury/contracts/src/ZrxTreasury.sol

```diff
@@ -0,0 +1,433 @@
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    /// @return proposalId The ID of the newly created proposal.
+    function propose(
+        ProposedAction[] memory actions,
+        uint256 executionEpoch,
+        string memory description,
+        bytes32[] memory operatedPoolIds
+    )
+        public
+        override
+        returns (uint256 proposalId)
+    {
+        require(
+            getVotingPower(msg.sender, operatedPoolIds) >= proposalThreshold,
+            "propose/INSUFFICIENT_VOTING_POWER"
+        );
+        require(
+            actions.length > 0,
+            "propose/NO_ACTIONS_PROPOSED"
+        );
+        uint256 currentEpoch = stakingProxy.currentEpoch();
+        require(
+            executionEpoch >= currentEpoch + 2,
+            "propose/INVALID_EXECUTION_EPOCH"
+        );
+
+        proposalId = proposalCount();
+        Proposal storage newProposal = proposals.push();
+
+        for (uint256 i = 0; i != actions.length; i++) {
+            newProposal.actions.push(actions[i]);
+        }
+        newProposal.executionEpoch = executionEpoch;
+        newProposal.voteEpoch = currentEpoch + 2;
+
+        emit ProposalCreated(
+            proposalId,
+            msg.sender,
+            description
+        );
+    }
+
+    /// @dev Casts a vote for the given proposal. Only callable
+    ///      during the voting period for that proposal. See
+    ///      `getVotingPower` for how voting power is computed.
+    /// @param proposalId The ID of the proposal to vote on.
+    /// @param support Whether to support the proposal or not.
+    /// @param operatedPoolIds The pools operated by `account`. The
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    function castVote(
+        uint256 proposalId,
+        bool support,
+        bytes32[] memory operatedPoolIds
+    )
+        public
+        override
+    {
+        if (proposalId >= proposalCount()) {
+            revert("castVote/INVALID_PROPOSAL_ID");
+        }
+        if (voteReceipts[proposalId][msg.sender].hasVoted) {
+            revert("castVote/ALREADY_VOTED");
+        }
+
+        Proposal memory proposal = proposals[proposalId];
+        if (
+            proposal.voteEpoch != stakingProxy.currentEpoch() ||
+            _hasVoteEnded(proposal.voteEpoch)
+        ) {
+            revert("castVote/VOTING_IS_CLOSED");
+        }
+
+        uint256 votingPower = getVotingPower(msg.sender, operatedPoolIds);
+        if (votingPower == 0) {
+            revert("castVote/NO_VOTING_POWER");
+        }
+
+        if (support) {
+            proposals[proposalId].votesFor = proposals[proposalId].votesFor
+                .safeAdd(votingPower);
+            voteReceipts[proposalId][msg.sender] = Receipt({
+                hasVoted: true,
+                support: true,
+                votingPower: votingPower
```

Comment:
why even store all this? UX? Can the frontend pull it from thegraph/events?

Reply:
yeah a lot of this stuff is an artifact of the Compound governance interface, I wasn't sure how much of this stuff the frontend needed. But based on the meeting this morning I'm going to delete a bunch of this